File: common/src/main/java/juuxel/adorn/block/KitchenSinkBlock.java
package juuxel.adorn.block;

import juuxel.adorn.block.entity.KitchenSinkBlockEntity;
import juuxel.adorn.util.ShapeRotation;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.InteractionHand;
import net.minecraft.world.InteractionResult;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.Items;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.EntityBlock;
import net.minecraft.world.level.block.entity.BlockEntity;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.phys.BlockHitResult;
import net.minecraft.world.phys.shapes.BooleanOp;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.phys.shapes.Shapes;
import net.minecraft.world.phys.shapes.VoxelShape;
import org.jetbrains.annotations.Nullable;

import java.util.EnumMap;
import java.util.Map;

public final class KitchenSinkBlock extends KitchenCounterBlock implements EntityBlock, BlockWithDescription {
    private static final Map<Direction, VoxelShape> SHAPES = new EnumMap<>(Direction.class);
    private static final String DESCRIPTION_KEY = "block.adorn.kitchen_sink.description";

    static {
        var sinkShapes = ShapeRotation.buildShapeRotationsFromNorth(
            2, 7, 3,
            14, 16, 14
        );
        AbstractKitchenCounterBlock.SHAPES.forEach((direction, shape) -> {
            SHAPES.put(direction, Shapes.join(shape, sinkShapes.get(direction), BooleanOp.ONLY_FIRST));
        });
    }

    public KitchenSinkBlock(Properties settings) {
        super(settings);
    }

    @Override
    public String getDescriptionKey() {
        return DESCRIPTION_KEY;
    }

    @Override
    protected InteractionResult useItemOn(ItemStack stack, BlockState state, Level world, BlockPos pos, Player player, InteractionHand hand, BlockHitResult hit) {
        var entity = world.getBlockEntity(pos, AdornBlockEntities.KITCHEN_SINK.get()).orElse(null);
        if (entity == null) return InteractionResult.TRY_WITH_EMPTY_HAND;
        boolean successful;

        if (stack.is(Items.SPONGE)) {
            successful = entity.clearFluidsWithSponge();

            if (successful) {
                stack.shrink(1);
                var wetStack = new ItemStack(Items.WET_SPONGE);

                if (stack.isEmpty()) {
                    player.setItemInHand(hand, wetStack);
                } else {
                    player.getInventory().placeItemBackInInventory(wetStack);
                }
            }
        } else {
            successful = entity.interactWithItem(stack, player, hand);
        }

        return successful ? InteractionResult.SUCCESS : InteractionResult.TRY_WITH_EMPTY_HAND;
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter view, BlockPos pos, CollisionContext context) {
        return SHAPES.get(state.getValue(FACING));
    }

    @Override
    public boolean hasAnalogOutputSignal(BlockState state) {
        return true;
    }

    @Override
    protected int getAnalogOutputSignal(BlockState state, Level world, BlockPos pos, Direction direction) {
        return world.getBlockEntity(pos) instanceof KitchenSinkBlockEntity sink ? sink.calculateComparatorOutput() : 0;
    }

    @Override
    public boolean isPathfindable(BlockState state, PathComputationType type) {
        return false;
    }

    @Override
    public @Nullable BlockEntity newBlockEntity(BlockPos pos, BlockState state) {
        return AdornBlockEntities.KITCHEN_SINK.get().create(pos, state);
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/KitchenSinkBlock.kt
package juuxel.adorn.block

import juuxel.adorn.block.variant.BlockVariant
import juuxel.adorn.util.buildShapeRotationsFromNorth
import net.minecraft.block.BlockEntityProvider
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.entity.BlockEntity
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.entity.player.PlayerEntity
import net.minecraft.item.ItemStack
import net.minecraft.item.Items
import net.minecraft.util.ActionResult
import net.minecraft.util.Hand
import net.minecraft.util.function.BooleanBiFunction
import net.minecraft.util.hit.BlockHitResult
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.util.shape.VoxelShapes
import net.minecraft.world.BlockView
import net.minecraft.world.World

class KitchenSinkBlock(variant: BlockVariant) : KitchenCounterBlock(variant), BlockEntityProvider, BlockWithDescription {
    override val descriptionKey = "block.adorn.kitchen_sink.description"

    override fun onUse(state: BlockState, world: World, pos: BlockPos, player: PlayerEntity, hand: Hand, hit: BlockHitResult): ActionResult {
        val entity = world.getBlockEntity(pos, AdornBlockEntities.KITCHEN_SINK).orElse(null) ?: return ActionResult.PASS
        val stack = player.getStackInHand(hand)
        val successful = if (stack.isOf(Items.SPONGE)) {
            val result = entity.clearFluidsWithSponge()

            if (result) {
                stack.decrement(1)
                val wetStack = ItemStack(Items.WET_SPONGE)

                if (stack.isEmpty) {
                    player.setStackInHand(hand, wetStack)
                } else {
                    player.inventory.offerOrDrop(wetStack)
                }
            }

            result
        } else {
            entity.interactWithItem(stack, player, hand)
        }

        return if (successful) ActionResult.success(world.isClient) else ActionResult.PASS
    }

    override fun getOutlineShape(state: BlockState, view: BlockView, pos: BlockPos, context: ShapeContext) =
        SHAPES[state[FACING]]

    override fun hasComparatorOutput(state: BlockState) = true

    override fun getComparatorOutput(state: BlockState, world: World, pos: BlockPos): Int =
        world.getBlockEntity(pos, AdornBlockEntities.KITCHEN_SINK).orElse(null)
            ?.calculateComparatorOutput() ?: 0

    override fun canPathfindThrough(state: BlockState, world: BlockView, pos: BlockPos, type: NavigationType) = false

    override fun createBlockEntity(pos: BlockPos, state: BlockState): BlockEntity? =
        AdornBlockEntities.KITCHEN_SINK.instantiate(pos, state)

    companion object {
        private val SHAPES: Map<Direction, VoxelShape>

        init {
            val sinkShapes = buildShapeRotationsFromNorth(
                2, 7, 3,
                14, 16, 14
            )
            SHAPES = AbstractKitchenCounterBlock.SHAPES.mapValues { (direction, shape) ->
                VoxelShapes.combineAndSimplify(
                    shape,
                    sinkShapes[direction] ?: error(
                        "Error building kitchen sink shapes: couldn't find sink shape for direction '$direction'"
                    ),
                    BooleanBiFunction.ONLY_FIRST
                )
            }
        }
    }
}
